mava/networks/attention.py

```diff
@@ -0,0 +1,93 @@
+# Copyright 2022 InstaDeep Ltd. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import chex
+import jax.numpy as jnp
+from flax import linen as nn
+from flax.linen.initializers import orthogonal
+
+
+class SelfAttention(nn.Module):
+    n_embd: int
+    n_head: int
+    n_agent: int
+    masked: bool = False
+
+    def setup(self) -> None:
+        # flax defualt is to only be defined using features out.
+        # bias init is zero in MAT.
+        assert self.n_embd % self.n_head == 0
+        self.key = nn.Dense(self.n_embd, kernel_init=orthogonal(0.01))
+        self.query = nn.Dense(self.n_embd, kernel_init=orthogonal(0.01))
+        self.value = nn.Dense(self.n_embd, kernel_init=orthogonal(0.01))
+
+        # output projection
+        self.proj = nn.Dense(self.n_embd, kernel_init=orthogonal(0.01))
+
+        # causal mask to ensure that attention is only applied to the left in the input sequence
+        self.mask = jnp.tril(jnp.ones((self.n_agent + 1, self.n_agent + 1))).reshape(
+            1, 1, self.n_agent + 1, self.n_agent + 1
+        )
```

Comment:
makes it a little clearer what we are doing I think (may need to be formatted for line length)
```suggestion
        self.mask = jnp.tril(jnp.ones((self.n_agent + 1, self.n_agent + 1)))[jnp.newaxis, jnp.newaxis]
```

Reply:
Done.